dataRetriever/dataPool/headersCache/headersNonceCache.go

```diff
@@ -0,0 +1,242 @@
+package headersCache
+
+import (
+	"bytes"
+	"github.com/ElrondNetwork/elrond-go/data"
+	"sync"
+	"time"
+)
+
+type headerListDetails struct {
+	headerList []headerDetails
+	timestamp  time.Time
+}
+
+type headerInfo struct {
+	headerNonce   uint64
+	headerShardId uint32
+}
+
+type headersNonceCache struct {
+	hdrNonceCache map[uint32]*headersMap
+
+	headersByHash *headersHashMap
+	hdrsCounter   *headersCounter
+
+	mutHeadersMap      sync.RWMutex
+	numHeadersToRemove int
+
+	maxHeadersPerShard int
+	canDoEviction      map[uint32]chan struct{}
+}
+
+func NewHeadersNonceCache(numHeadersToRemove int, numMaxHeaderPerShard int) *headersNonceCache {
+	return &headersNonceCache{
+		hdrNonceCache:      make(map[uint32]*headersMap),
+		hdrsCounter:        newHeadersCounter(),
+		headersByHash:      newHeadersHashMap(),
+		mutHeadersMap:      sync.RWMutex{},
+		numHeadersToRemove: numHeadersToRemove,
+		canDoEviction:      make(map[uint32]chan struct{}),
+		maxHeadersPerShard: numMaxHeaderPerShard,
+	}
+}
+
+func (hnc *headersNonceCache) addHeaderInNonceCache(headerHash []byte, header data.HeaderHandler) bool {
+
+	headerShardId := header.GetShardID()
+	headerNonce := header.GetNonce()
+
+	// add header info in second map
+	headerInfo := headerInfo{
+		headerNonce:   header.GetNonce(),
+		headerShardId: header.GetShardID(),
+	}
+	alreadyExits := hnc.headersByHash.addElement(headerHash, headerInfo)
+	if alreadyExits {
+		return true
+	}
+
+	headerShardPool := hnc.getShardMap(headerShardId)
+
+	//critical section
```

Comment:
delete comments from inside the function.

Reply:
done.